src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationTestFixture.cs

```diff
@@ -34,68 +43,130 @@ public class WebApplicationTestFixture<TStartup> : IDisposable where TStartup :
-        /// </para>
-        /// </summary>
-        /// <param name="solutionSearchPattern">The glob pattern to use when searching for a solution file by
-        /// traversing up the folder hierarchy from the test execution folder.</param>
-        /// <param name="solutionRelativePath">The path to the project folder relative to the solution file of your
-        /// application. The folder of the first sln file that matches the <paramref name="solutionSearchPattern"/>
-        /// found traversing up the folder hierarchy from the test execution folder is considered as the base path.</param>
-        protected WebApplicationTestFixture(string solutionSearchPattern, string solutionRelativePath)
+        private void EnsureServer()
         {
+            if (_server != null)
+            {
+                return;
+            }
+
+            EnsureBuilder();
             EnsureDepsFile();
 
-            var builder = CreateWebHostBuilder();
-            builder
-                .UseStartup<TStartup>()
-                .UseSolutionRelativeContentRoot(solutionRelativePath);
+            SetContentRoot();
+
+            ConfigureWebHost(_builder);
+            _server = CreateServer(_builder);
+        }
+
+        private void SetContentRoot()
+        {
+            var metadataAttributes = GetContentRootMetadataAttributes(typeof(TStartup).Assembly.FullName);
+            string contentRoot = null;
+            for (var i = 0; i < metadataAttributes.Length; i++)
+            {
+                var contentRootAttribute = metadataAttributes[i];
+                var contentRootCandidate = Path.Combine(
+                    AppContext.BaseDirectory,
+                    contentRootAttribute.ContentRootPath);
+
+                var contentRootMarker = Path.Combine(
+                    contentRootCandidate,
+                    Path.GetFileName(contentRootAttribute.ContentRootTest));
 
-            ConfigureWebHost(builder);
-            _server = CreateServer(builder);
+                if (File.Exists(contentRootMarker))
+                {
+                    contentRoot = contentRootCandidate;
+                    break;
+                }
+            }
 
-            Client = CreateClient();
+            if (contentRoot != null)
+            {
+                _builder.UseContentRoot(contentRoot);
+            }
+            else
+            {
+                _builder.UseSolutionRelativeContentRoot(typeof(TStartup).Assembly.GetName().Name);
+            }
+        }
+
+        private static WebApplicationFixtureContentRootAttribute[] GetContentRootMetadataAttributes(string key)
+        {
+            var testAssembly = GetCandidateAssemblies();
+            var metadataAttributes = testAssembly
+                .SelectMany(a => a.GetCustomAttributes<WebApplicationFixtureContentRootAttribute>())
+                .Where(a => string.Equals(a.Key, key, StringComparison.OrdinalIgnoreCase))
+                .OrderBy(a => int.Parse(a.Priority))
```

Comment:
I asked earlier if this should be an `int` to begin with. But if not, this needs to pass in invariant culture. And I'm also worried about this throwing, so it should guard against that using TryParse and maybe assuming `0` if it can't parse.